Assets/_Application/Layers/LayerTypes/Polygons/PolygonInputToLayer.cs

```diff
@@ -173,56 +191,67 @@ public void ShowPolygonVisualisations(bool enabled)
         {
             foreach (var polygonLayer in layers.Values)
             {
-                if (polygonLayer.IsMask)
+                PolygonSelectionLayerPropertyData propertyData = polygonLayer.GetProperty<PolygonSelectionLayerPropertyData>();
+                if (propertyData.IsMask)
                 {
                     continue;
                 }
-
-                polygonLayer.PolygonVisualisation.gameObject.SetActive(enabled);
+                var key = layers.FirstOrDefault(x => x.Value == polygonLayer).Key;
+                key.gameObject.SetActive(enabled);
             }
 
             selectionCalculator.gameObject.SetActive(enabled);
         }
 
         private void CreatePolygonLayer(List<Vector3> unityPolygon)
         {
-            _ = new PolygonSelectionLayer(
-                "Polygon", 
-                polygonSelectionVisualisationPrefab.PrefabIdentifier, 
-                unityPolygon, 
-                ShapeType.Polygon,
-                onSpawn: data =>
+            ILayerBuilder layerBuilder = LayerBuilder.Create()
+                .OfType(polygonSelectionVisualisationPrefab.PrefabIdentifier)
+                .NamedAs("Polygon")
+                .AddProperty(new PolygonSelectionLayerPropertyData
                 {
-                    if (data is not PolygonSelectionLayer layer) return;
-
-                    layers.Add(layer.PolygonVisualisation, layer);
-                    layer.polygonSelected.AddListener(ProcessPolygonSelection);
-                    polygonInput.SetDrawMode(PolygonInput.DrawMode.Edit); //set the mode to edit explicitly, so the reselect functionality of ProcessPolygonSelection() will immediately work
-                }
-            );            
+                    ShapeType = ShapeType.Polygon,
+                    OriginalPolygon = unityPolygon.ToCoordinates().ToList()
+                });
+            App.Layers.Add(layerBuilder, visualisation =>
+            {
+                PolygonSelectionVisualisation polygonVisualisation = visualisation as PolygonSelectionVisualisation;
+                layers.Add(polygonVisualisation, visualisation.LayerData);
+                PolygonSelectionLayerPropertyData data = visualisation.LayerData.GetProperty<PolygonSelectionLayerPropertyData>();
+                data.polygonSelected.AddListener(ProcessPolygonSelection);
+                polygonInput.SetDrawMode(PolygonInput.DrawMode.Edit);
+                ProcessPolygonSelection(visualisation.LayerData);
+            });
         }
 
         private void UpdateLayer(List<Vector3> editedPolygon)
         {
-            ActiveLayer.SetShape(editedPolygon.ToCoordinates().ToList());
+            //TODO is this the right way to get the polygon visualisation layergameobject?
```

Comment:
This TODO pops up a lot and seems important?

Reply:
zie boven